registry/cache/memcached.go

```diff
@@ -124,60 +118,29 @@ func NewFixedServerMemcacheClient(config MemcacheConfig, addresses ...string) Cl
 	return newClient
 }
 
-func (c *memcacheClient) GetKey(k Keyer) ([]byte, error) {
+func (c *memcacheClient) GetKey(k Keyer) ([]byte, time.Time, error) {
 	cacheItem, err := c.Get(k.Key())
 	if err != nil {
 		if err == memcache.ErrCacheMiss {
 			// Don't log on cache miss
-			return []byte{}, ErrNotCached
+			return []byte{}, time.Time{}, ErrNotCached
 		} else {
 			c.logger.Log("err", errors.Wrap(err, "Fetching tag from memcache"))
-			return []byte{}, err
+			return []byte{}, time.Time{}, err
 		}
 	}
-	var data expiryData
-	err = json.Unmarshal(cacheItem.Value, &data)
-	if err != nil {
-		return []byte{}, err
-	}
-	return data.Data, nil
-}
-
-// GetExpiration returns the expiry time of the key
-func (c *memcacheClient) GetExpiration(k Keyer) (time.Time, error) {
-	cacheItem, err := c.Get(k.Key())
-	if err != nil {
-		if err == memcache.ErrCacheMiss {
-			// Don't log on cache miss
-			return time.Time{}, ErrNotCached
-		} else {
-			c.logger.Log("err", errors.Wrap(err, "Fetching tag from memcache"))
-			return time.Time{}, err
-		}
-	}
-
-	var data expiryData
-	err = json.Unmarshal(cacheItem.Value, &data)
-	if err != nil {
-		return time.Time{}, err
-	}
-	return time.Unix(int64(data.Expiry), 0), nil
+	expiry := binary.BigEndian.Uint32(cacheItem.Value)
+	return cacheItem.Value[4:], time.Unix(int64(expiry), 0), nil
 }
 
 func (c *memcacheClient) SetKey(k Keyer, v []byte) error {
-	expiry := int32(time.Now().Add(expiry).Unix())
-	data := expiryData{
-		Expiry: expiry,
-		Data:   v,
-	}
-	val, err := json.Marshal(&data)
-	if err != nil {
-		return err
-	}
+	expiry := time.Now().Add(expiry).Unix()
+	expiryBytes := make([]byte, 4, 4)
+	binary.BigEndian.PutUint32(expiryBytes, uint32(expiry))
 	if err := c.Set(&memcache.Item{
 		Key:        k.Key(),
-		Value:      val,
-		Expiration: expiry,
+		Value:      append(expiryBytes, v...),
+		Expiration: int32(expiry),
```

Comment:
We set the expiry for memcached so that old items are garbage-collected; but since the memcached client doesn't report its own record of the expiry back to us, we have to store it in the item value as well.

It's not the best scheme, to be honest. But it deserves a bit of explanation, until we change it, at least :)

Reply:
I see what you mean. I missed the fact the data structure with the Expiration field is not a flux data structure but something that an external dependency uses. :+1: